### PR TITLE
Add Progress Bar Component

### LINKED ISSUE #1365: Add Progress Bar component
Probably a similar API to `ProgressRing`.

https://www.lightningdesignsystem.com/components/progress-bar/

### PR DESCRIPTION
**Pull request by aswinshenoy**
Fixes #1365 

Added Progress Bar Component support.

@interactivellama @futuremint, kindly review it, I will make any changes required.

### Additional description
The progress bar component as found in SLDS has been ported to react, with almost all its functionality, and reusing the same classes.

#### Props Taken
* ```id``` - HTML id for the component container
* ```className``` - CSS classes to be added to tag with `.slds-progress-bar`.
* ```label``` - Label for the progress bar, if provided will render as a descriptive progress bar
* ```color``` - Color to be filled in the progress bar
* ```thickness``` - Thickness of the progress bar
* ```value``` - Percentage of progress completion, ranging [0, 100].

#### Limitations
* Vertical Progress bar has not been implemented as of now, but can easily be added
* Color, Radius are limited to already defined classes in SLDS.

#### Screenshots
Default - 
image
With Label -
image
Color set to success -
image
Different Thickness -
image
Rounded -
image

---

### CONTRIBUTOR checklist (do not remove)
Please complete for every pull request

* [x] First-time contributors should sign the Contributor License Agreement. It's a fancy way of saying that you are giving away your contribution to this project. If you haven't before, wait a few minutes and a bot will comment on this pull request with instructions.
* [x] `npm run lint:fix` has been run and linting passes.
* [x] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [x] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Mocha, Jest (Storyshots), and `components/component-docs.json` CI tests pass (`npm test`).
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).

### REVIEWER checklist (do not remove)

* [ ] TravisCI tests pass. This includes linting, Mocha, Jest, Storyshots, and `components/component-docs.json` tests.
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] The Accessibility panel of each Storybook story has 0 violations (aXe). Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).
###### Required only if there are markup / UX changes
* [ ] Add year-first date and commit SHA to `last-slds-markup-review` in `package.json` and push.
* [ ] Request a review of the deployed Heroku app by the Salesforce UX Accessibility Team.
* [ ] Add year-first review date, and commit SHA, `last-accessibility-review`, to `package.json` and push.
* [ ] While the contributor's branch is checked out, run `npm run local-update` within locally cloned [site repo](https://github.com/salesforce-ux/design-system-react-site) to confirm the site will function correctly at the next release.

### CODE REVIEW

**Comment by aswinshenoy** [comment]
I learning and working on adding test cases, as per given instructions in the contribution guide. I will update with a patch, adding it in sometime soon. :)

**Comment by interactivellama** [comment]
@aswinshenoy Screenshots are looking great! I'll try to get you feedback on your props/API proposal soon.

**Comment by aswinshenoy** [comment]
@interactivellama meanwhile, I will add the component into ```components/component-docs.json```, and also try writing tests.
Are multiple commits per PR okay?

Regards,

**Comment by interactivellama** [comment]
yes, multiple commits are fine.

File: components/progress-bar/index.jsx [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	}
+
+	render() {
+		return (
+			<div id={this.getId()}>
+				{this.getDescription()}
+				<div
+					className={classNames(
+						'slds-progress-bar',
+						this.getRadius(),
```
Comment (tanhengyeow) [inline components/progress-bar/index.jsx]: Actually, you can try:
```
{  `slds-progress-bar_${this.props.radius}`: this.props.radius }
```
That way you don't need to declare a dedicated function to return an empty string in the event this props wasn't passed in.

File: components/progress-bar/index.jsx [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+						this.getThickness(),
```
Comment (tanhengyeow) [inline components/progress-bar/index.jsx]: Same here. If you just use:
```
`slds-progress-bar_${this.props.thickness}`
```
You won't even need the function.

File: components/progress-bar/index.jsx [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+						this.props.className
+					)}
+					style={{
+						width: this.progressPercent(),
```
Comment (tanhengyeow) [inline components/progress-bar/index.jsx]: Can't we just do it inline here instead of using a function?

File: components/progress-bar/index.jsx [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+					}}
+				>
+					<span
+						className={classNames('slds-progress-bar__value', this.getColor())}
```
Comment (tanhengyeow) [inline components/progress-bar/index.jsx]: Same here.

File: components/progress-bar/index.jsx (hunk 1 above)
Comment (aswinshenoy) [reply components/progress-bar/index.jsx]: :+1:

File: components/progress-bar/index.jsx (hunk 2 above)
Comment (aswinshenoy) [reply components/progress-bar/index.jsx]: :+1:

File: components/progress-bar/index.jsx (hunk 4 above)
Comment (aswinshenoy) [reply components/progress-bar/index.jsx]: :+1:

File: components/progress-bar/index.jsx (hunk 3 above)
Comment (aswinshenoy) [reply components/progress-bar/index.jsx]: :+1:

File: components/progress-bar/index.jsx (hunk 1 above)
Comment (interactivellama) [reply components/progress-bar/index.jsx]: Please refer to https://github.com/salesforce/design-system-react/blob/master/docs/codebase-overview.md#use-entire-class-name-string-when-rendering-conditional-class-names

File: components/progress-bar/index.jsx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	 */
+	className: PropTypes.oneOfType([
+		PropTypes.array,
+		PropTypes.object,
+		PropTypes.string,
+	]),
+	/**
+	 * Label for the progress bar
+	 */
+	label: PropTypes.string,
```
Comment (interactivellama) [inline components/progress-bar/index.jsx]: Please allow node

**Comment by interactivellama** [comment]
Props proposal is approved. Please ping when you are ready for a review.

File: components/progress-bar/index.jsx (hunk 5 above)
Comment (aswinshenoy) [reply components/progress-bar/index.jsx]: @interactivellama  I have updated it into `node` type

**Comment by interactivellama** [comment]
Props approved. Please proceed with tests, since this isn't interactive DOM and image snapshots should be enough.

@aswinshenoy The prop type for `label` should be able to be node and string.

**Comment by interactivellama** [comment]
Also, you have lint errors

**Comment by aswinshenoy** [comment]
I have added snapshot test, and written `progress-bar.snapshot-test.jsx` similar to that of `progress-ring.snapshot-test.jsx`.

**Comment by interactivellama** [comment]
@aswinshenoy Because I was unclear about how the Jest snapshot tests were to be created, I went ahead and pushed an update to your branch. https://github.com/salesforce/design-system-react/pull/1773/commits/9dcb6d5f5948199dfc74c9463d831453efb15e8c#diff-ef6040eafe786a4a9723ca820a5e2566R43

The `id` prop in the examples need to be set so that the DOM snapshots are always the same when rendered.

**Comment by aswinshenoy** [comment]
@interactivellama I have pushed adding ids in the examples, and the tests have passed it seems :) Sorry, I was a bit late and inactive for some days.

**Comment by interactivellama** [comment]
@aswinshenoy I will be the _last_ person to complain about being late and inactive (at least before and after the official GSoC time 😄 ).

**Comment by interactivellama** [comment]
As an aside, @aswinshenoy it's great that you have added some of the building block components that will be consumed by Setup Assistant.

File: components/progress-bar/index.jsx [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	getDescription() {
+		if (this.props.label) {
+			return (
+				<div
+					className="slds-grid slds-grid_align-spread slds-p-bottom_x-small"
+					id="progress-bar-label-id-4"
+				>
+					<span>{this.props.label}</span>
+					<span aria-hidden="true">
+						<strong>{this.props.value}% Complete</strong>
```
[the remark this answers is not in the document]
Comment (aswinshenoy) [reply components/progress-bar/index.jsx]: True, I forgot about the internationalization part. Seems like you already fixed it with `labels` prop. Thank you :)

**Comment by interactivellama** [comment]
Looks good! Thank you @aswinshenoy

File: components/progress-bar/index.jsx [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+					className={classNames(
+						'slds-progress-bar',
+						this.props.radius ? `slds-progress-bar_${this.props.radius}` : null,
+						this.props.thickness
+							? `slds-progress-bar_${this.props.thickness}`
+							: null,
+						this.props.className
+					)}
+					style={{
+						width: `${this.props.value}%`,
```
[the remark this answers is not in the document]
Comment (aswinshenoy) [reply components/progress-bar/index.jsx]: Sorry, I must have been a little careless there. :(  Looks perfect with your patch :+1:

**Comment by aswinshenoy** [comment]
@interactivellama It is wonderful to see it done :) I will looking forward for more :+1:  Thank you so much for writing out the test, and doing all those changes. :100:  I had been a little sloppy on this, perhaps, sorry.

**Comment by interactivellama** [comment]
I just talked to a program manager today on an internal team, and they are looking to consume this and ScopedNotification, so I wanted to go ahead and get this these merged.

**Comment by aswinshenoy** [comment]
@interactivellama that's cool, thank you. The #1738 PR that adds ScopedNotifications also has also passed the tests and seems it is mergeable now after your fixes.

**Pull request merged**